packages/app/src/runner/event-manager.ts

```diff
@@ -603,6 +603,27 @@ export class EventManager {
       Cypress.primaryOriginCommunicator.toAllSpecBridges('before:unload')
     })
 
+    Cypress.on('request:snapshot:from:spec:bridge', ({ log, name, options, specBridge, addSnapshot }: {
+      log: Cypress.Log
+      name?: string
+      options?: any
+      specBridge: string
+      addSnapshot: (snapshot: any, options: any, shouldRebindSnapshotFn: boolean) => Cypress.Log
+    }) => {
+      const eventID = log.get('id')
+
+      Cypress.primaryOriginCommunicator.once(`snapshot:for:log:generated:${eventID}`, (generatedCrossOriginSnapshot) => {
+        const snapshot = generatedCrossOriginSnapshot.body ? generatedCrossOriginSnapshot : null
+
+        addSnapshot.apply(log, [snapshot, options, false])
+      })
+
+      Cypress.primaryOriginCommunicator.toSpecBridge(specBridge, 'generate:snapshot:for:log', {
+        name,
+        id: eventID,
+      })
```

Comment:
This might be one of those cases where adding a promise-based API to the communicator would be useful. Then it could be something like:

```js
const generatedCrossOriginSnapshot = await 
Cypress.primaryOriginCommunicator.toSpecBridgeAsync(specBridge, 'generate:snapshot:for:log')

const snapshot = generatedCrossOriginSnapshot.body ? generatedCrossOriginSnapshot : null

addSnapshot.apply(log, [snapshot, options, false])
```

I'm not suggesting adding it in this PR, but wanted to call it out as an example. Maybe when we have another use-case, we could add that and refactor this one.